File: math/number_theory/eratosthenes_sieve/python/eratosthenes_sieve.py
from typing import List

def eratosthenes_sieve(n: int) -> List[int]:
    if n < 2:
        return []
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    for i in range(2, int(n ** 0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
    return [i for i in range(2, n + 1) if is_prime[i]]

if __name__ == "__main__":
    assert eratosthenes_sieve(0) == []
    assert eratosthenes_sieve(1) == []
    assert eratosthenes_sieve(10) == [2, 3, 5, 7]
    primes_20 = eratosthenes_sieve(20)
    assert len(primes_20) == 8 and primes_20[-1] == 19
    assert len(eratosthenes_sieve(100)) == 25
    print("Eratosthenes sieve tests passed.")

File: math/number_theory/eratosthenes_sieve/cpp/eratosthenes_sieve.cpp
#include <vector>
#include <cassert>
#include <iostream>

std::vector<int> eratosthenesSieve(int n) {
    std::vector<bool> isPrime(n+1, true);
    isPrime[0] = isPrime[1] = false;
    for (int i = 2; i * i <= n; ++i) {
        if (isPrime[i]) {
            for (int j = i * i; j <= n; j += i) {
                isPrime[j] = false;
            }
        }
    }
    std::vector<int> primes;
    for (int i = 2; i <=n; ++i) {
        if (isPrime[i]) {
            primes.push_back(i);
        }
    }
    return primes;
}

int main() {

    assert(eratosthenesSieve(0).empty());
    assert(eratosthenesSieve(1).empty());

    std::vector<int> expected = {2, 3, 5, 7};
    assert(eratosthenesSieve(10) == expected);

    std::vector<int> primes_20 = eratosthenesSieve(20);
    assert(primes_20.size() == 8);
    assert(primes_20.back() == 19);

    assert(eratosthenesSieve(100).size() == 25);
    std::cout << "Eratosthenes sieve tests passed.\n";
    return 0;
}